File: 2018/02/02.py
from itertools import product

def generator(input): 
    return input.splitlines()

def part_1(input):
    twice = 0
    thrice = 0
    for id in input: 
        counts = [id.count(c) for c in id]
        twice += 2 in counts
        thrice += 3 in counts
    return twice * thrice        
        
def part_2(input):
    for ida, idb in product(input, input):
        if sum([ida[i] != idb[i] for i in range(len(ida))]) == 1 : 
            return ''.join([item for item in ida if item in idb])
        

File: tmpevents/2018/02/02.rs
aoc::main!();

fn preprocessing(input: &str) -> Vec<&str> {
    input
    .lines()
    .collect_vec()
}

fn part_1(box_ids: Vec<&str>) -> usize {
    box_ids
    .iter()
    .map(|box_id| box_id.chars().counts().into_values().collect_vec())
    .map(|cnts| (cnts.contains(&2) as usize, cnts.contains(&3) as usize))
    .fold((0, 0, 0), |acc, cnt| (acc.0 + cnt.0, acc.1 + cnt.1, (acc.0 + cnt.0) * (acc.1 + cnt.1)))
    .2
}

fn part_2(box_ids: Vec<&str>) -> String {
    iproduct!(box_ids.clone(), box_ids)
    .map(|(ida, idb)| 
        zip(ida.chars(), idb.chars())
        .filter(|(a, b)| a == b)
        .map(|(a, _)| a)
        .join(""))
    .find(|ida| ida.len() == 25)
    .unwrap()
}